autode/opt/optimisers/robust.py

```diff
@@ -0,0 +1,458 @@
+        *args,
+        **kwargs,
+    ):
+        """
+        Hybrid RFO and TRIM/QA optimiser with dynamic trust radius (and
+        optional damping if oscillation in energy and gradient is detected)
+
+        ---------------------------------------------------------------------
+        Args:
+            coord_type: 'DIC' for delocalised internal coordinates, 'cart'
+                               for Cartesian coordinates
+            init_trust: Initial value of trust radius in Angstrom
+            update_trust: Whether to update the trust radius or not
+            min_trust: Minimum bound for trust radius (only for trust update)
+            max_trust: Maximum bound for trust radius (only for trust update)
+            damp: Whether to apply damping if oscillation is detected
+            *args: Additional arguments for ``NDOptimiser``
+            **kwargs: Additional keyword arguments for ``NDOptimiser``
+
+        Keyword Args:
+            maxiter (int): Maximum number of iterations (from ``NDOptimiser``)
+            gtol (GradientRMS): Tolerance on RMS(|∇E|) (from ``NDOptimiser``)
+            etol (PotentialEnergy): Tolerance on |E_i+1 - E_i|
+                                    (from ``NDOptimiser``)
+
+        See Also:
+            :py:meth:`NDOptimiser <NDOptimiser.__init__>`
+        """
+        super().__init__(*args, init_alpha=init_trust, **kwargs)
+
+        assert 0.0 < min_trust < max_trust
+        self._max_alpha = float(max_trust)
+        self._min_alpha = float(min_trust)
+        self._upd_alpha = bool(update_trust)
+
+        if coord_type.lower() in ["cart", "cartesian"]:
+            self._coord_type = "cart"
+        elif coord_type.lower() == "dic":
+            self._coord_type = "dic"
+        else:
+            raise ValueError("Coordinate type must be either 'cart' or 'DIC'")
+        # todo stop if any constraints detected -> is this needed
+        self._should_damp = bool(damp)
+        self._last_damp_iter = 0
+
+        self._hessian_update_types = [FlowchartUpdate]
+
+    @property
+    def _g_norm(self) -> GradientRMS:
+        """Calculate the RMS gradient norm in Cartesian coordinates"""
+        grad = self._coords.to("cart").g
+        return GradientRMS(np.sqrt(np.average(np.square(grad))))
+
+    @property
+    def converged(self) -> bool:
+        if self._species is not None and self._species.n_atoms == 1:
+            return True  # Optimisation 0 DOF is always converged
+
+        # gradient is better indicator of stationary point
+        if self._g_norm < self.gtol / 10:
+            logger.warning(
+                f"Gradient norm criteria overachieved. "
+                f"{self._gtol.to('Ha/ang')/10:.3f} Ha/Å. "
+                f"Signalling convergence"
+            )
+
+        # both gradient and energy must be converged!!
+        return self._abs_delta_e < self.etol and self._g_norm < self.gtol
+
+    def _initialise_run(self) -> None:
+        self._build_coordinates()
```

Comment:
if constraints aren't supported then 
```suggestion
    def _initialise_run(self) -> None:
        assert not self._species.constraints.any
        self._build_coordinates()
```
?

Reply:
@shoubhikraj I am not quite sure how the species constraints work. If I define cartesian or distance constraint in species.constraints, does that reflect in the calculated gradient/hessians? If so, then it would work without complaint with this optimsier. If not, it would probably fail.